<launch>
  <!-- Login credentials and request parameters -->
  <arg name="base_url" default="$(optenv INTEROP_HOST http://interop:80)"/>
  <arg name="username" default="$(optenv INTEROP_USERNAME testadmin)"/>
  <arg name="password" default="$(optenv INTEROP_PASSWORD testpass)"/>
  <arg name="timeout" default="1.0"/>

  <!-- Targets directory settings -->
  <arg name="targets_root" default="~/object_files/"/>
  <arg name="interop_update_period" default="10.0"/>

  <!-- Synchronization settings -->
  <arg name="sync_queue_size" default="2"/>
  <arg name="max_sync_delay" default="1"/>

  <!-- Subscribed topics -->
  <arg name="navsat_topic" default="/mavros/global_position/global"/>
  <arg name="compass_topic" default="/mavros/global_position/compass_hdg"/>

  <!-- Published topics -->
  <arg name="moving_topic" default="~moving"/>
  <arg name="stationary_topic" default="~stationary"/>
  <arg name="flyzones_topic" default="~flyzones" />
  <arg name="search_grid_topic" default="~search_grid"/>
  <arg name="waypoints_topic" default="~waypoints"/>
  <arg name="air_drop_topic" default="~air_drop_loc"/>
  <arg name="emergent_targ_topic" default="~emergent_targ_loc"/>
  <arg name="off_axis_targ_topic" default="~off_axis_targ"/>
  <arg name="utm_zone_topic" default="~utm_zone"/>

  <!-- Publication periods -->
  <arg name="obstacles_period" default="0.05"/>
  <arg name="mission_info_period" default="1.00"/>

  <!-- Frame IDs -->
  <arg name="obstacles_frame" default="odom"/>
  <arg name="missions_frame" default="odom"/>

  <!-- Mission ID -->
  <arg name="mission_id" default="-1"/>

  <group ns="interop">
    <!-- Obstacles client -->
    <node name="obstacles"
          pkg="interop"
          type="obstacles_client.py"
          output="screen"
          respawn="true">
      <!-- Login credentials and request parameters -->
      <param name="base_url" value="$(arg base_url)"/>
      <param name="username" value="$(arg username)"/>
      <param name="password" value="$(arg password)"/>
      <param name="timeout" value="$(arg timeout)"/>

      <!-- Published topics -->
      <param name="moving_topic" value="$(arg moving_topic)"/>
      <param name="stationary_topic" value="$(arg stationary_topic)"/>

      <!-- Publication period -->
      <param name="period" value="$(arg obstacles_period)"/>

      <!-- Frame ID -->
      <param name="frame" value="$(arg obstacles_frame)"/>
    </node>

    <!-- Mission Information Client -->
    <node name="mission_info"
          pkg="interop"
          type="missions_client.py"
          output="screen"
          respawn="true">
      <!-- Login credentials and request parameters -->
      <param name="base_url" value="$(arg base_url)"/>
      <param name="username" value="$(arg username)"/>
      <param name="password" value="$(arg password)"/>
      <param name="timeout" value="$(arg timeout)"/>

      <!-- Published topics -->
      <param name="flyzones_topic" value="$(arg flyzones_topic)"/>
      <param name="search_grid_topic" value="$(arg search_grid_topic)"/>
      <param name="waypoints_topic" value="$(arg waypoints_topic)"/>
      <param name="air_drop_loc_topic" value="$(arg air_drop_topic)"/>
      <param name="emergent_targ_topic" value="$(arg emergent_targ_topic)"/>
      <param name="off_axis_targ_topic" value="$(arg off_axis_targ_topic)"/>
      <param name="utm_zone_topic" value="$(arg utm_zone_topic)"/>

      <!-- Publication period -->
      <param name="period" value="$(arg mission_info_period)"/>

      <!-- Frame ID -->
      <param name="frame" value="$(arg missions_frame)"/>

      <!-- Mission ID -->
      <param name="id" value="$(arg mission_id)"/>
    </node>

    <!-- Telemetry client -->
    <node name="telemetry"
          pkg="interop"
          type="telemetry_client.py"
          output="screen"
          respawn="true">
      <!-- Login credentials and request parameters -->
      <param name="base_url" value="$(arg base_url)"/>
      <param name="username" value="$(arg username)"/>
      <param name="password" value="$(arg password)"/>
      <param name="timeout" value="$(arg timeout)"/>

      <!-- Synchronization settings -->
      <param name="sync_queue_size" value="$(arg sync_queue_size)"/>
      <param name="max_sync_delay" value="$(arg max_sync_delay)"/>

      <!-- Subscribed topics -->
      <param name="navsat_topic" value="$(arg navsat_topic)"/>
      <param name="compass_topic" value="$(arg compass_topic)"/>
    </node>

    <!-- Target client -->
    <node name="targets"
          pkg="interop"
          type="targets_server.py"
          output="screen"
          respawn="true">
      <!-- Login credentials and request parameters -->
      <param name="base_url" value="$(arg base_url)"/>
      <param name="username" value="$(arg username)"/>
      <param name="password" value="$(arg password)"/>
      <param name="timeout" value="$(arg timeout)"/>

      <!-- Targets directory settings -->
      <param name="targets_root" value="$(arg targets_root)"/>
      <param name="interop_update_period" value="$(arg interop_update_period)"/>
    </node>
  </group>
</launch>
